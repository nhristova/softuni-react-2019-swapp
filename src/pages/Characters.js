import React from 'react';
import { useQuery } from '@apollo/react-hooks';
import gql from 'graphql-tag.macro';

export const CHARACTERS_QUERY = gql`
  query AllPeople($perPage: Int!) {
    allPeople(first: $perPage) 
    {
      edges{
        node {
          id
          name
          image
        }
      }
    }
  }
`;

export const Characters = () => {
  const { data, loading, error } = useQuery(CHARACTERS_QUERY, {
    variables: { perPage: 12 }
  });

  if (loading) return <div>Loading</div>;
  if (error) return <p>Error</p>;

  const {
    allPeople: { edges }
  } = data;

  return (
    <main className="characters-main">
      <div className="characters-list">
        {edges.map(ed => <CharacterLRCard character={ed.node} key={ed.node.id} />)}
      </div>
      <button className="load-more-button">Load more</button>
    </main>
  );
};

export const CharacterLRCard = (prop) => {
  const character = prop.character;
  return (
    <div className="small-lr-card lr-card">
      <img
        className="lr-card-photo"
        src={character.image}
        alt={character.name}
      />
      <div className="lr-card-details">
        <div className="lr-card-title">
          <a
            className="heading-starwars text-highlight "
            href={`./characters/${character.id}`}
          >
            {character.name}
              </a>
        </div>
      </div>
    </div>
  )
}
