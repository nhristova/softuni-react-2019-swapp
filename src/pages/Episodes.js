import React from 'react';
import gql from 'graphql-tag.macro';
import { useQuery } from '@apollo/react-hooks';

export const EPISODES_QUERY = gql`
  query{
    allEpisodes(first: 100){
      edges{
        node{
          id
          title
          episodeId
          image    
          openingCrawl
        }
      }
    }
  }
`;

export const Episodes = () => {
  const { data, loading, error } = useQuery(EPISODES_QUERY);
  if (loading) return <div>Loading</div>;
  if (error) return <p>Error on getting all cars</p>;

  const {
    allEpisodes: { edges }
  } = data;

  return (
    <main className="episodes-main">
      {edges.map(ed => <EpisodeTBCard episode={ed.node} key={Math.random()} />)}
    </main>
  );
};

export const EpisodeTBCard = (prop) => {
  const episode = prop.episode;
  return (
    <div className="episode-card tb-card">
      <div className="tb-card-photo">
        <img
          src={episode.image}
          alt={`Poster ${episode.title}`}
        />
      </div>
      <div className="tb-card-title">
        <a className="heading-starwars text-highlight " href="./episode">{episode.title}</a>
      </div>
      <div className="tb-card-description">{episode.openingCrawl}</div>
    </div>
  );
}
