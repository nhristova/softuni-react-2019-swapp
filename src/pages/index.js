export * from './Character';
export * from './Characters';
export * from './Episode';
export * from './Episodes';
export * from './Starship';
export * from './Login';
export * from './Home';
export * from './Routes';
