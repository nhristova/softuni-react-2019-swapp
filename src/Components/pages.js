import React from 'react';
import { Switch, Redirect, Route } from 'react-router-dom';
import { Header } from './Header';
import { Episodes } from './Episodes';
import { Characters } from './Characters';

export const Pages = props => {
  return (
    <>
      <Header changeTheme={props.toggleTheme} />
      <Switch>
        <Route path="/episodes" component={Episodes} />
        <Route path="/characters" component={Characters} />
        <Redirect from="/login" to="/episodes" />
        <Redirect from="/" to="/episodes" />
      </Switch>
    </>
  );
};
