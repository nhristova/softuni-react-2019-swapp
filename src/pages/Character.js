import React from 'react';
import gql from 'graphql-tag.macro';
import { useQuery } from '@apollo/react-hooks';
import { Detail, LRCard } from '../Components';
import { useParams } from 'react-router-dom';

export const CHARACTER_QUERY = gql`
  query($characterId: ID!) {
    person(id: $characterId) {
      id
      name
      birthYear
      height
      mass
      image
      homeworld {
        id
        name
      }
      species {
        id
        name
      }
      starships(first: 100) {
        edges {
          node {
            id
            name
            image
          }
        }
      }
    }
  }
`;

export const Character = props => {
  const { characterId } = useParams();
  const { data, loading, error } = useQuery(CHARACTER_QUERY, {
    variables: { characterId: characterId },
  });

  if (loading) return <div>Loading</div>;
  if (error) return <p>Error</p>;

  const { person } = data;

  return (
    <main className="character-main">
      <div className="page-title page-title-1 heading-starwars text-highlight text-highlight ">
        {person.name}
      </div>
      <div className="page-content">
        <div className="page-column-half big-lr-card tb-card">
          <div className="tb-card-title heading-starwars text-highlight ">
            {person.name}
          </div>
          <div className="tb-card-photo">
            <img src={person.image} alt={person.name} />
          </div>
          <div className="tb-card-details">
            <Detail type="Height" value={person.height} />
            <Detail type="Weight" value={person.mass} />
            <Detail type="Species" value={person.species.name} />
            <Detail type="Home World" value={person.homeworld.name} />
          </div>
        </div>
        <div className="page-column-half starhips-list">
          <div className="column-title heading-starwars text-highlight2 page-title-2">
            Piloted Starships
          </div>
          {person.starships.edges.map(ship => (
            <LRCard
              item={ship.node}
              key={ship.node.id}
              page="starships"
              size="tiny-lr-card"
            />
          ))}
        </div>
      </div>
    </main>
  );
};
