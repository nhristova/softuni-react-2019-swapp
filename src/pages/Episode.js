import React from 'react';
import gql from 'graphql-tag.macro';
import { useQuery } from '@apollo/react-hooks';
import { Detail, LRCard } from '../Components';
import { useParams } from 'react-router-dom';

export const EPISODE_QUERY = gql`
  query($episodeId: ID!, $perPage: Int, $after: String) {
    episode(id: $episodeId) {
      id
      title
      image
      openingCrawl
      director
      releaseDate
      people(first: $perPage, after: $after) {
        pageInfo {
          hasNextPage
          endCursor
        }
        edges {
          node {
            id
            name
            image
          }
        }
      }
    }
  }
`;

export const Episode = props => {
  const { episodeId } = useParams();
  const { data, loading, error } = useQuery(EPISODE_QUERY, {
    variables: {
      episodeId: episodeId,
      perPage: 5,
    },
  });

  if (loading) return <div>Loading</div>;
  if (error) return <p>Error</p>;

  const { episode } = data;

  // const loadMore = () => {
  //   fetchMore({
  //     variables: {
  //       perPage: 2,
  //       after: episode.people.pageInfo.endCursor
  //     },
  //     updateQuery: (prev, { fetchMoreResult: { episode: episodeNew } }) => {
  //       return {
  //         episode: {
  //           ...episodeNew,
  //           people: {
  //             t
  //             pageInfo: {
  //               ...episodeNew.people.pageInfo
  //             },
  //             edges: [
  //               ...prev.episode.people.edges,
  //               ...episodeNew.people.edges
  //             ]
  //           }
  //         }
  //       }
  //     }
  //   })
  // }

  return (
    <main className="episode-main">
      <div className="big-lr-card lr-card">
        <img
          className="lr-card-photo"
          src={episode.image}
          alt={episode.title}
          width="100%"
        />
        <div className="lr-card-details">
          <div className="lr-card-title heading-starwars text-highlight ">
            {episode.title}
          </div>
          <div className="lr-card-subtitle heading-starwars text-highlight-2">
            subtitle??
          </div>
        </div>
      </div>
      <div className="episode-details">
        <div className="episode-synopsis">{episode.openingCrawl}</div>
        <div className="episode-info">
          <Detail type="Director" value={episode.director} />
          <Detail type="Release date" value={episode.releaseDate} />
        </div>
      </div>
      <div className="characters-list">
        {episode.people.edges.map(ch => (
          <LRCard
            item={ch.node}
            key={ch.node.id}
            size="small-lr-card"
            page="characters"
          />
        ))}
      </div>
      {episode.people.pageInfo.hasNextPage && (
        <button className="load-more-button" onClick={() => {}}>
          Load more
        </button>
      )}
    </main>
  );
};
