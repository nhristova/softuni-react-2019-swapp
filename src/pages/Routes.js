import React from 'react';
import { Switch, Redirect, Route } from 'react-router-dom';
import { Header } from '../Components';
import { Episodes } from './Episodes';
import { Characters } from './Characters';

export const Routes = props => {
  return (
    <>
      <Header changeTheme={props.toggleTheme} />
      <Switch>
        <Route path="/episodes" component={Episodes} />
        <Route path="/characters" component={Characters} />
        <Redirect from="/login" to="/episodes" />
        <Redirect from="/" to="/episodes" />
      </Switch>
    </>
  );
};
