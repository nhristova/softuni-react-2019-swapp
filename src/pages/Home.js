import React, { useState, useEffect } from 'react';
import gql from 'graphql-tag.macro';
import { useQuery } from '@apollo/react-hooks';
import { Routes } from './Routes';
import { Login } from './Login';
import { Redirect } from 'react-router-dom';

export const AUTHENTICATED_QUERY = gql`
  query IsAuthenticated {
    authenticated @client
  }
`;

/* istanbul ignore next*/
export const Home = () => {
  const { data } = useQuery(AUTHENTICATED_QUERY);

  let [theme, setTheme] = useState('light-theme');

  const toggleTheme = () => {
    const newTheme = theme === 'light-theme' ? 'dark-theme' : 'light-theme';
    setTheme(newTheme);
  };

  useEffect(() => {
    // TODO: Get theme from cookie? Use Context to store theme
    document.body.classList = theme;
  }, [theme]);

  return data.authenticated ? (
    <Routes toggleTheme={toggleTheme} />
  ) : (
    <>
      <Login toggleTheme={toggleTheme} />
      <Redirect to="/login" />
    </>
  );
};
