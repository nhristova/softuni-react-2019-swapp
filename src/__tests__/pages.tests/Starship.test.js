import React from 'react';
import renderer from 'react-test-renderer';
import { MockedProvider, wait } from '@apollo/react-testing';
import { Starship, STARSHIP_QUERY } from '../../pages';
import { MemoryRouter } from 'react-router-dom';

describe('Starship page', () => {
  it('Renders loader', async () => {
    let result = renderer.create(
      <MemoryRouter initialEntries={['/starships/starships.12']}>
        <MockedProvider mocks={[]}>
          <Starship />
        </MockedProvider>
        ,
      </MemoryRouter>,
    );

    expect(result).toMatchSnapshot();

    await renderer.act(async () => {
      await wait(0);
    });
  });

  it.skip('Renders mock data', async () => {
    const mock = {
      request: {
        query: STARSHIP_QUERY,
        variables: { perPage: 12 },
      },
      result: {
        data: {
          starship: {
            id: 'starships.12',
            name: 'X-wing',
            model: 'T-65 X-wing',
            image:
              'https://static.turbosquid.com/Preview/2015/10/14__02_29_23/xwingtopleft_01_open_01.jpgb5dc9c7c-25bc-44f8-88ba-50e41873111aOriginal.jpg',
            starshipClass: 'starfighter',
            cost: 149999,
            maxAtmosphericSpeed: 1050,
            maxMLPerHour: 100,
            hyperdriveRating: 1,
            crew: 1,
          },
        },
      },
    };

    let result = renderer.create(
      <MemoryRouter initialEntries={['/starships/starships.12']}>
        <MockedProvider mocks={[mock]} addTypename={false}>
          <Starship />
        </MockedProvider>
      </MemoryRouter>,
    );

    await renderer.act(async () => {
      await wait(0);
    });

    expect(result).toMatchSnapshot();
  });
});
