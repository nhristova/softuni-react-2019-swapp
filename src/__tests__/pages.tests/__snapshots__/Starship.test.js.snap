// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Starship page Renders loader 1`] = `
Array [
  <div>
    Loading
  </div>,
  ",",
]
`;

exports[`Starship page Renders mock data 1`] = `
Array [
  <p>
    Data
  </p>,
  ",",
]
`;
